### PR TITLE
feat:add upload download add bandwidth limit

### PR DESCRIPTION
**Pull request by constwz**
### Description

Upload and download with broadband stream implementation

### Rationale

Protect the sp service

### Example

Limit the upload and download speed

### Changes

Notable changes: 
* Upload and download with broadband stream implementation

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ type StorageProviderConfig struct {
 	RateLimiter        *localhttp.RateLimiterConfig
 	DiscontinueCfg     *stopserving.DiscontinueConfig
 	MetadataCfg        *metadata.MetadataConfig
+	BandwidthLimiter   *localhttp.BandWidthLimiterConfig
```
Comment (will-2012) [inline config/config.go:47]: Rename `BandWidthLimiterConfig ` to `BandwidthLimiterConfig `.

File: config/config.go [hunk 2]
```diff
@@ -208,6 +210,10 @@ var DefaultRateLimiterConfig = &localhttp.RateLimiterConfig{
 	},
 }
 
+var DefaultBandwidthLimiterConfig = &localhttp.BandWidthLimiterConfig{
+	Enable: false,
```
Comment (will-2012) [inline config/config.go:214]: Maybe you need to update the default configuration synchronously at config/config_template.toml.

File: pkg/middleware/http/tokenBucket.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	"sync"
+
+	"github.com/bnb-chain/greenfield-storage-provider/pkg/log"
+	"golang.org/x/time/rate"
+)
+
+type BandWidthLimiterConfig struct {
+	Enable bool
+	R      rate.Limit
+	B      int
```
Comment (will-2012) [inline pkg/middleware/http/tokenBucket.go:13]: Add more commentaries to explain the config meaning. 
This naming convention is a bit vague for the name that the user needs to configure, and the user may not understand it.

File: service/gateway/object_handler.go [hunk 4]
```diff
@@ -117,7 +118,12 @@ func (gateway *Gateway) getObjectHandler(w http.ResponseWriter, r *http.Request)
 			errDescription = makeErrorDescription(merrors.GRPCErrorToInnerError(err))
 			return
 		}
-
+		// If it is limited, it will block
+		if localHttp.BandWidthLimit != nil {
+			if err := localHttp.BandWidthLimit.Limiter.Wait(ctx); err != nil {
+				log.Errorw("failed to Wait bandwidth limiter", "err", err)
```
Comment (will-2012) [inline service/gateway/object_handler.go:124]: Refine `Wait` to `wait`, `err` to `error` in log.

File: service/gateway/object_handler.go [hunk 5]
```diff
@@ -223,6 +229,12 @@ func (gateway *Gateway) putObjectHandler(w http.ResponseWriter, r *http.Request)
 			return
 		}
 		if readN > 0 {
+			// If it is limited, it will block
+			if localHttp.BandWidthLimit != nil {
+				if err := localHttp.BandWidthLimit.Limiter.Wait(ctx); err != nil {
```
Comment (will-2012) [inline service/gateway/object_handler.go:234]: I don't understand how to calculate the actual bandwidth of the service?

File: config/config.go (hunk 1 above)
Comment (constwz) [reply config/config.go]: fixed

File: config/config.go (hunk 2 above)
Comment (constwz) [reply config/config.go:214]: added

File: pkg/middleware/http/tokenBucket.go (hunk 3 above)
Comment (constwz) [reply pkg/middleware/http/tokenBucket.go]: Annotated

**Review by krish-nr — APPROVED** [review]
LGTM

**Review by joeylichang — APPROVED** [review]

**Pull request merged**